### PR TITLE
top p search and testing

### PR DESCRIPTION
**Pull request by jessechancy**

### CODE REVIEW

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Thanks! Mainly looks good, left some comments

File: keras_nlp/utils/text_generation.py [hunk 1]
```diff
@@ -393,3 +393,146 @@ def token_probability_fn(inputs):
...
+        cumulative_probs = tf.math.cumsum(sorted_preds, axis=-1)
+        # Create a mask for the tokens to keep.
+        keep_mask = cumulative_probs <= p
+        # Shift to include the last token that exceed p.
+        shifted_keep_mask = tf.concat(
+            [tf.ones_like(keep_mask[:, :1]), keep_mask[:, :-1]], axis=-1
+        )
+        # Filter out unmasked tokens and sample from filtered distribution.
+        probs = tf.where(
+            shifted_keep_mask, sorted_preds, tf.fill(pred.shape, filter_value)
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:519]: Yep filter_value should be 0, edited

File: keras_nlp/utils/text_generation.py [hunk 2]
```diff
@@ -393,3 +393,146 @@ def token_probability_fn(inputs):
...
+        from_logits: bool. Indicates whether `token_probability_fn` outputs
+            logits or probabilities.
+        end_token_id: int, defaults to None. The token marking the end of the
+            sequence, once encountered the generation is finished for the exact
+            sequence. If None, every sequence is generated up to `max_length`.
+            If set, all tokens after encountering `end_token_id` will be
+            replaced with `pad_token_id`.
+        pad_token_id: int, defaults to 0. The pad token after `end_token_id`
+            is received.
+        filter_value: float, defaults to -Inf. The value for filtering out
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:436]: I don't think we need to allow customization, we can just mask out tokens that are unused by having filter_value = 0

File: keras_nlp/utils/text_generation.py [hunk 3]
```diff
@@ -393,3 +393,146 @@ def token_probability_fn(inputs):
...
+
+    i = prompt.shape[1]
+    while i < max_length:
+        # If the prompt has reached our desired length, exit while loop.
+        pred = token_probability_fn(prompt)
+        if from_logits:
+            pred = tf.keras.activations.softmax(pred, axis=-1)
+        # Sort preds in descending order.
+        sorted_preds, sorted_indices = tf.math.top_k(
+            pred, k=pred.shape[-1], sorted=True
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation.py:507]: edited

File: keras_nlp/utils/text_generation_test.py [hunk 4]
```diff
@@ -418,3 +419,175 @@ def token_probability_fn(inputs):
             seed=42,
         )
         self.assertAllEqual(output_logit, output_probs)
+
+
+class TopPSamplingTextGenerationTest(tf.test.TestCase):
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply keras_nlp/utils/text_generation_test.py:424]: edited for this and other testing names

File: keras_nlp/utils/text_generation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply keras_nlp/utils/text_generation.py:435]: Yea, in that case, let's remove this argument.

File: keras_nlp/utils/text_generation_test.py [hunk 5]
```diff
@@ -418,3 +419,175 @@ def token_probability_fn(inputs):
...
+            rtol=0.2,
+        )
+
+    def test_only_choose_from_top_p_tokens(self):
+        def token_probability_fn(inputs):
+            batch_size = inputs.shape[0]
+            prob = tf.constant([[0.4, 0.3, 0.2, 0.1]])
+            return tf.repeat(prob, batch_size, axis=0)
+
+        # Test that it only samples from tokens that sum up to p
```
Comment (chenmoneygithub) [inline keras_nlp/utils/text_generation_test.py:528]: nit: period at end.

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
only a few minor comments!

**Review by chenmoneygithub — APPROVED** [review]

File: keras_nlp/utils/text_generation.py [hunk 6]
```diff
@@ -393,3 +393,145 @@ def token_probability_fn(inputs):
...
+    p,
+    seed=None,
+    from_logits=False,
+    end_token_id=None,
+    pad_token_id=0,
+):
+    """
+    Text generation utility based on top-p (nucleus) sampling.
+
+    Top-p search filters the top token with probabilities that sum up to p, and
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:411]: maybe we can explain this a little more clearly, also remember to backtick argument names so it's clear what they are...

Top-p search selects tokens from the smallest subset of output probabilities that sum to greater than `p`. Put another way, top-p will first order token predictions by likelihood, and ignore all tokens after the cumulative probably of selected tokens exceeds `p`.

**Review by mattdangerw — APPROVED** [review]
This looks great! Nice work.

Just a few minor comments

File: keras_nlp/utils/text_generation.py [hunk 7]
```diff
@@ -112,7 +116,7 @@ def token_probability_fn(inputs):
 
     prompt = tf.fill((BATCH_SIZE, 1), START_ID)
 
-    # Print the generated sequence (token ids).
```
Comment (mattdangerw) [inline keras_nlp/utils/text_generation.py:115]: typo?

Comment (mattdangerw) [reply keras_nlp/utils/text_generation.py]: fixed
